### PR TITLE
feat: Logging key metrics for web search

### PR DESCRIPTION
**Pull request by seanssc-hue**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) to discuss your idea/fix with the community before creating a pull request, and describe your changes before submitting a pull request.

This is to ensure large feature PRs are discussed with the community first, before starting work on it. If the community does not want this feature or it is not relevant for Open WebUI as a project, it can be identified in the discussion before working on the feature and submitting the PR.

**Before submitting, make sure you've checked the following:**

- [ ] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [ ] **Description:** Provide a concise description of the changes made in this pull request down below.
- [ ] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
- [ ] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [ ] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [ ] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [ ] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [ ] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [ ] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [ ] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- Added new module for dispatched web searches that logs key metrics such as query, engine, user, latency, etc...

### Added

- Added `web_search_logger` module that logs at three different times: before search being dispatched to the upstream provider, after the search provider returns and following search results retrieval. These combination of logging includes key metrics such as query, engine, urls fetched, user info, latency, etc...

### Changed

- [List any changes, updates, refactorings, or optimizations]

### Deprecated

- [List any deprecated functionality or features that have been removed]

### Removed

- [List any removed features, files, or functionalities]

### Fixed

- [List any fixes, corrections, or bug fixes]

### Security

- [List any new or updated security-related changes, including vulnerability fixes]

### Breaking Changes

- **BREAKING CHANGE**: [List any breaking changes affecting compatibility or functionality]

---

### Additional Information

- [Insert any additional context, notes, or explanations for the changes]
  - [Reference any related issues, commits, or other relevant information]

### Screenshots or Videos

image
image

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

**Review by pascal-beaulne — COMMENTED** [review]
I"m not 100% sure on the implementation here. Maybe using a context manager + a dataclass. we've also added opentelemetry. maybe we could leverage that as well.

File: backend/open_webui/routers/retrieval.py [hunk 1]
```diff
@@ -1463,6 +1522,15 @@ def load_docs_incremental() -> list[Document]:
...
             except Exception as e:
+                content_load_ms = round((time.monotonic() - content_load_start) * 1000)
+                log_web_search_content_load(
+                    event_id=audit_event_id,
+                    urls=urls,
+                    docs_loaded=0,
+                    latency_ms=content_load_ms,
+                    engine=request.app.state.config.RAG_WEB_SEARCH_ENGINE,
+                    error=e,
+                )
```
[the remark this answers is not in the document]
Comment (seanssc-hue) [reply backend/open_webui/routers/retrieval.py:1535]: @copilot apply changes based on [this feedback](https://github.com/ssc-dsai/canchat-v2/pull/481#discussion_r3052952208)

File: backend/open_webui/retrieval/web/web_search_logger.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+            "timestamp": ts,
+            "engine": engine,
+            "query": query,
+            "query_hash": _hash_query(query),
+            "query_length": len(query),
+            "result_count_requested": result_count_requested,
+            "domain_filter_active": bool(domain_filter_list),
+            "domain_filter_count": len(domain_filter_list) if domain_filter_list else 0,
+            **_safe_user_attrs(user),
+        }
```
[the remark this answers is not in the document]
Comment (seanssc-hue) [reply backend/open_webui/retrieval/web/web_search_logger.py:261]: @copilot review

File: backend/open_webui/retrieval/web/web_search_logger.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        record = {
+            "event_id": event_id,
+            "event_type": "web_search_result",
+            "timestamp": datetime.now(timezone.utc).isoformat(),
+            "engine": engine,
+            "latency_ms": elapsed_ms,
+            "result_count": len(results) if results else 0,
+            "result_urls": result_urls,
+            "result_domains": result_domains,
+            "result_domain_count": len(result_domains),
```
[the remark this answers is not in the document]
Comment (seanssc-hue) [reply backend/open_webui/retrieval/web/web_search_logger.py:332]: @copilot review

File: backend/open_webui/retrieval/web/web_search_logger.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+            ),
+        )
+    except Exception:
+        log.exception("[web_search_audit] failed to initialize OTel metrics")
+        return WebSearchMetricsInstruments()
+
+
+_METRICS_INSTRUMENTS = _create_metrics_instruments()
+_MAX_LOGGED_RESULT_URLS = 20
+_MAX_LOGGED_CONTENT_LOAD_URLS = 20
```
Comment (pascal-beaulne) [inline backend/open_webui/retrieval/web/web_search_logger.py]: Doesn't seemed to be used.

**Review by pascal-beaulne — APPROVED** [review]
